### PR TITLE
Feature: Tolerant flight checks for docked ships

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature:** There is no PR discussing this issue (that I could find).
## Feature Details
This PR makes the takeoff process for docked (carried) ships more realistic by lifting the requirements for power generation and engines on those ships.

Although it is not strictly necessary, this PR should probably wait for #6726.

**Rationalization:**

Previously, all ships had to have power generation and engines installed in order to be able to take off from a planet. However, if a ship is inside a carrier, it doesn't actually need any engines to take off - the carrier will just take it up into space. In this PR, the thurster, steering and power generation requirements are lifted from such ships.

We also know that docked ships are connected to the mothership's power systems, since their batteries can be replenished. Since they are inside a dock, they have life support from the mothership, so even if they don't have power the pilot wouldn't suffocate.

Realistically, there is no difference between a cargo container in the cargo bay and between a stripped down Boxwing in a fighter bay. In fact, this is pretty much the essence of a Boxwing.

**Implementation Details:**

Flight checks are only used for the player's ships, and is done via the `PlayerInfo::FlightCheck` method. This method ties into the planetary takeoff process and the outfitter/shop tooltips, so changing the behaviour here should be enough to fully support this feature (in theory, anyway).

`Ship::FlightCheck` now has a parameter for docked ships; if it is false, the previous behaviour is preserved. If it is true, the power, thurster and steering errors are downgraded to warnings.

`PlayerInfo::FlightCheck` now stores which ships are docked, instead of only counting the number of docks and docked ships. This required some minor code flow changes:
```
check flight issues and count docks
attempt to dock ships
```
got changed to
```
count docks
attempt to dock ships
check flight issues
```

**Notes:**
- This feature relies on the predictability of ship autodocking (on takeoff). If that was changed in a nondeterministic way, this feature would become unviable. Such a change could affect gameplay in a major way, and is therefore not probable.
- Reordering ships can make a previously viable fleet unable to take off; in this case, the warnings automatically change to errors in the side panel.
- One could argue that the current balance for Boxwing (and other fighters/drones) is based on the requirements for power and engines, and ships without these could become overly powerful. This is not the case, since those fighters would be stationary, and would get killed almost instantly in a battle.
- It was already possible to use (dock and deploy) such fighters in the game, however they couldn't depart from a planet. One can create such fighters by taking all their power and engines after they have been disabled.
- These fighters can be deployed by themselves and do receive the standard launch velocity, however they cannot dock unless the mothership manually manouvers to them.
- Feel free to edit the new tooltips, I'm sure someone more professional can make them a bit more natural and understandable.

## UI Screenshots

Energy warning for docked ships:
Screenshot_20220809_201937
Energy warning for non-docked ships (same as existing behaviour):
Screenshot_20220809_202139

## Testing Done
Playtesting has been done with multiple fighters and a carrier where some of the fighters had missing components.

## Performance Impact
This feature shouldn't have a significant performance impact, considering that the original bay counting already ran in n^2 time, and the extra iteration through ships in the new version only adds n (where n is the number of ships in the player's fleet).

### CODE REVIEW

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -833,10 +836,10 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
 				{
 					Files::LogError("Expected bay to be empty for " + ship->ModelName() + ": " + ship->Name());
 					categoryCount[bay.ship->Attributes().Category()].emplace_back(bay.ship);
+					dockedShips[bay.ship]=true;
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:839]:
```suggestion
					dockedShips[bay.ship] = true;
```

File: source/PlayerInfo.cpp [hunk 2]
```diff
@@ -851,20 +854,28 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
...
-				// Include the lack of bay availability amongst any other
-				// warnings for this carriable ship.
-				auto it = flightChecks.find(carriable);
-				string warning = "no bays?";
-				if(it != flightChecks.end())
-					it->second.emplace_back(warning);
-				else
-					flightChecks.emplace(carriable, vector<string>{warning});
+				--bayType.second;
+				dockedShips[carriable]=true;
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:859]:
```suggestion
				dockedShips[carriable] = true;
```

File: source/PlayerInfo.cpp [hunk 3]
```diff
@@ -851,20 +854,28 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
...
 			}
+			else
+				dockedShips[carriable]=false;
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:862]:
```suggestion
				dockedShips[carriable] = false;
```

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -851,20 +854,28 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
...
 		}
 	}
+
+	for(const auto &ship : ships)
+		if(ship->GetSystem() == system && !ship->IsDisabled() && !ship->IsParked())
+		{
+			// Check if ship is docked
+			auto it = dockedShips.find(ship);
+			auto checks = ship->FlightCheck(it!=dockedShips.end() && it->second);
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:871]:
```suggestion
			auto checks = ship->FlightCheck(it != dockedShips.end() && it->second);
```

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -851,20 +854,28 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
...
+			// Include the lack of bay availability amongst any other
+			// warnings for this carriable ship.
+			if(it!=dockedShips.end() && !it->second)
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:874]:
```suggestion
			if(it != dockedShips.end() && !it->second)
```

File: source/Ship.cpp [hunk 6]
```diff
@@ -1104,21 +1105,27 @@ vector<string> Ship::FlightCheck() const
 	// Report the first error condition that will prevent takeoff:
 	if(IdleHeat() >= MaximumHeat())
 		checks.emplace_back("overheating!");
-	else if(energy <= 0.)
+	else if(energy <= 0. && !docked)
```
Comment (Hurleveur) [inline source/Ship.cpp:1108]: We can only check for docked once, and then keep the logic as it was before, in this intended block.
Like this:
```cpp
	else if(!docked)
	{
		if(energy <= 0.)
			checks.emplace_back("no energy!");
		else if((energy - consuming <= 0.) && (fuel <= 0.))
			checks.emplace_back("no fuel!");
		else if(!thrust && !reverseThrust && !afterburner)
			checks.emplace_back("no thruster!");
		else if(!turn)
			checks.emplace_back("no steering!");
	}
```

File: source/Ship.cpp [hunk 7]
```diff
@@ -1104,21 +1105,27 @@ vector<string> Ship::FlightCheck() const
...
 	// If no errors were found, check all warning conditions:
 	if(checks.empty())
 	{
-		if(!thrust && !reverseThrust)
+		if(energy <= 0.)
+			checks.emplace_back("no energy?");
+		if(!thrust && !afterburner && !reverseThrust)
+			checks.emplace_back("no thruster?");
+		if(!turn)
+			checks.emplace_back("no steering?");
```
Comment (Hurleveur) [inline source/Ship.cpp:1125]: We could use a warning for energy - consuming <= 0. I think. (the no fuel! from above had that check)
Or is it not included for a specific reason?

Also we can optimize this a bit by only checking for these added conditions if the ship is a docked ship.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looks good to me, no big issues with the code, and the ability to completely convert those ships into cargo is a feature I've wanted too.
Just some small convention problems (variable = something, with spaces) you can apply the batch by going into the files changed tab and clicking add to batch until you've got them all.

File: source/Ship.cpp (hunk 7 above)
Comment (tibetiroka) [reply source/Ship.cpp:1130]: I've added a new check for `(energy - consuming <= 0.) && (fuel <= 0.)` which matches the previous version. Not sure if you wanted this or just a general warning about ships using fuel generators only; that could be added after the `fuel?` check.

Comment (Hurleveur) [reply source/Ship.cpp:1130]: I would have been fine with either, you can mark this as resolved

File: source/PlayerInfo.cpp [hunk 8]
```diff
@@ -808,19 +808,22 @@ const vector<shared_ptr<Ship>> &PlayerInfo::Ships() const
 // Returns a mapping of ships to the reason their flight check failed.
 map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
 {
+	// Stores the relations between ships and docks.
+	// True: The ship is docked
+	// False: The ship requires a dock but is not docked
+	// Missing: The ship doesn't require a dock, and is therefore not docked
+	auto dockedShips = map<shared_ptr<Ship>,bool>{};
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:815]: Best this way, so we don't initialise one to transfer it later.

Also no need to initialize it as empty it does it by itself.
```suggestion
	map<shared_ptr<Ship>, bool> dockedShips;
```

File: source/PlayerInfo.cpp [hunk 9]
```diff
@@ -851,20 +854,28 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
...
+		{
+			// Check if ship is docked
+			auto it = dockedShips.find(ship);
+			auto checks = ship->FlightCheck(it != dockedShips.end() && it->second);
+			// Include the lack of bay availability amongst any other
+			// warnings for this carriable ship.
+			if(it != dockedShips.end() && !it->second)
+				checks.emplace_back("no bays?");
+			if(!checks.empty())
+				flightChecks.emplace(ship, checks);
```
Comment (Hurleveur) [inline source/PlayerInfo.cpp:877]: May as well save it in a bool and name it (more for clarity than for optimization, really)
```suggestion
			const bool isCarried = it != dockedShips.end();
			auto checks = ship->FlightCheck(isCarried && it->second);
			// Include the lack of bay availability amongst any other
			// warnings for this carriable ship.
			if(isCarried && !it->second)
				checks.emplace_back("no bays?");
			if(!checks.empty())
				flightChecks.emplace(ship, checks);
```

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looking good! Few more things I noticed could be improved, and then someone probably should indeed go over the tips.
I did not actually playtest this, that is probably not required though, and I assume you did.

**Comment by tibetiroka** [comment]
All done. I playtested most things so it should be fine. As soon as the new tooltips are reviewed this should be ready for merging.

**Review by Hurleveur — APPROVED** [review]
I hope its ready for merging indeed, seems good to me at least, the only thing that bothers me a bit is the fact we pretty much test the same thing twice (no energy etc) except one is with if and the other one with else if but I don't know how to fix that, probably not possible really.

**Comment by tibetiroka** [comment]
Yeah, that and the use of `dockedShips` is a bit wonky, but I don't think there is a better way without refactoring the whole thing.

**Comment by Hurleveur** [comment]
@roadrunner56 could you take a look a those new descriptions when you have the time?
they seem fine to me but that doesn't mean much

File: data/tooltips.txt [hunk 10]
```diff
@@ -938,12 +938,24 @@ tip "no thruster!"
 tip "no steering!"
 	`This ship does not have any steering systems installed. If it took off, it would be unable to turn.`
 
+tip "no energy?"
+	`This ship does not have any source of energy to power its systems. Although it can take off while docked on an other ship, it will not function once deployed.`
```
Comment (roadrunner56) [inline data/tooltips.txt:942]:
```suggestion
	`This ship does not have any source of energy to power its systems. Although it can take off while docked on another ship, it will not function once deployed.`
```

File: data/tooltips.txt [hunk 11]
```diff
@@ -938,12 +938,24 @@ tip "no thruster!"
...
+	`This ship does not have enough fuel to provide energy to power its systems. Although it can take off while docked on an other ship, it will not function once deployed.`
+
 tip "afterburner only?"
 	`This ship's only thruster is an afterburner. It may be unable to move if the afterburner runs out of fuel.`
 
 tip "reverse only?"
 	`This ship has a reverse thruster but no forward thruster, which could make it very difficult to maneuver.`
 
+tip "no thruster?"
+	`This ship does not have any sort of thruster installed. Although it can take off while docked on an other ship, it would be unable to move once deployed.`
```
Comment (roadrunner56) [inline data/tooltips.txt:954]:
```suggestion
	`This ship does not have any sort of thruster installed. Although it can take off while docked on another ship, it will be unable to move once deployed.`
```

**Review by roadrunner56 — CHANGES_REQUESTED** [review]

**Comment by tibetiroka** [comment]
Thanks! Apparently I wasn't paying much attention to the details there.

**Review by roadrunner56 — APPROVED** [review]

**Comment by petervdmeer** [comment]
I'm not sure if this change improves the game. Some players will appreciate being able to fly around with incomplete, broken, fighters, but this seems to be a bit of niche functionality.

Please first check with @Amazinite and @Zitchas if this change is desired before I generate a lot of additional work for this PR.

And if they agree that this is a good addition to the game, then I will ask for extensive test-coverage for this feature (integration tests with things like "fighters that have fuel and hyperdrives, but no energy", in various orders with other complete and incomplete fighters).
Not wanting to discourage anyone, but niche functionality needs quite extensive (automated) testing, especially since it will not be used often (and typically forgotten by manual tests in future PRs).

I will then also request that we make sure that such "incomplete, broken," fighters will not be deployed by the deploy command (but that a message is given instead that says something like "[nr] ships were not deployed because they lacked the means to function independently.").
Such an additional change will help with minimizing the support requests that we can expect from players that are unaware of the functionality in this PR and that accidentally take off with an incomplete fighter.

**Comment by tibetiroka** [comment]
> the support requests that we can expect from players that are unaware of the functionality in this PR and that accidentally take off with an incomplete fighter.

I'm not sure why this feature would cause trouble by itself; the behaviour of AI fleets have not changed and players get warnings in the outfitter and in the shipyard about having such ships. The only way you could avoid those warnings is if you captured these ships, in which case the existing behaviour would be even worse (since it is inconsistent).

> this seems to be a bit of niche functionality

At first glance, maybe. However, this PR enables a lot more than an extra cargo expansion on your Boxwings. Consider this: stripped down ships are cheap, customizable and expendable.

They are cheap, because depending on the type of ship you buy, the outfits might be 14-88% of the cost. (The most extreme examples are the Boxwing, Black Diamond and Violin Spider; other ships fall closer to the 30-60% mark.) This allows you to easily assemble huge (although severely handicapped) fleets.

They are customizable, since you can put just about anything in them, unless you want to use them as fighters - in which case, this feature is indeed irrelevant. You can use them to store ammo, which could be useful if the mothership uses missile launchers.

Also, don't forget that you can remove _any_ of the previously required generators / engies, but yon don't need to remove _all_ of them. You could make ships that have power and reverse thrusters, but no steering - unless they use weapons, they will be able to stop, and even return to a stationary mothership (assuming the AI can figure out reverse-only thrushers). And if you install long-range missile launchers on those ships they become deployable super-mines that are relatively cheap and can beat anything that doesn't shoot back. They can be especially useful against single targets.

Due to their low cost, these ships are also expendable. I haven't playtested this yet, but I can imagine that a fleeing ship could release these fighters to distract its pursuers, presenting an alternative to dumping cargo. Okay, this _is_ a niche use case, but I could totally see this being used merchant fleets.

And these are only the things off the top of my head. Considering that @Hurleveur wanted similar changes for a while there will be even more potential uses and demand for this feature.

> fighters that have fuel and hyperdrives, but no energy

Hyperdrives, Scram Drives and Jump Drives don't actually consume energy when used. And, again, any "new" edge cases have already been present in the game, even if they don't have bug reports / tests associated with them. If you want these tests, they have to be added even if this PR never gets merged.

> I will then also request that we make sure that such "incomplete, broken," fighters will not be deployed by the deploy command

That could be discussed; I see your point here but deploying these ships _could_ be useful. I just don't see a way where we could allow deploying these ships without including them in the default command. If that were possible, sure thing. Maybe we could change how the deploy command selects docked ships (using the existing flight checks), I'll have to look into that.

TL;DR: I don't think this is a niche use case, and any issues this PR allegedly introduces are already existing edge cases in the game.

**Comment by Hurleveur** [comment]
> That could be discussed; I see your point here but deploying these ships could be useful. I just don't see a way where we could allow deploying these ships without including them in the default command. If that were possible, sure thing. Maybe we could change how the deploy command selects docked ships (using the existing flight checks), I'll have to look into that.

We could make it a preference, or a warning when you deploy ships.

---------------------------------

Also there is no need to be worried by integration tests, they don't require too much work to understand/do, and they assure everyone that this thing works.
In this case, 4 simple tests with save files that contain the different kind of previously prevented from launching fighter ships, as well as one with these limitations on normal ships.
You just make a new procedure to attempt launching, that will press "D" then "Enter", and then "L" and check the date. If the date changed then you indeed launched, otherwise you had a warning.

But maybe you should not get too much work going before Amazinite approves this behaviour.

**Comment by petervdmeer** [comment]
> I will then also request that we make sure that such "incomplete, broken," fighters will not be deployed by the deploy command (but that a message is given instead that says something like "[nr] ships were not deployed because they lacked the means to function independently.").

@tibetiroka  Please consider to create this other PR first, because this other PR will also be useful for many other situations (and I don't expect many objections to this other PR).

**Comment by Pilcrow182** [comment]
> We could make it a preference, or a warning when you deploy ships. (you have unequipped fighters, do you want to launch them as well? ok/cancel dialog)

Might I suggest both of these, actually? A warning dialogue with a yes/no prompt, but also a preference that, when enabled, skips that dialogue and proceeds as if the player had selected 'yes'.

**Comment by tibetiroka** [comment]
Alright, here's how I'm planning on implementing the new preferences.
I'll create separate preferences for the takeoff and the deployment behaviours. I'm not a big fan of that, but everything else just seems ambiguous (feel free to post suggestions though). The deployment preference will be in a separate PR with most of the tests, while the takeoff preference will be included here.

The main issue is that I want `always`, `ask` and `never` options for takeoff (defaults to `ask` for clarity), but only `always` and `never` for deploying them, since adding popup prompts to combat scenarios seems suboptimal.

**Comment by tibetiroka** [comment]
Alright, here are the requested options. I added two of them, which is probably too much, so please tell me your opinion below.

The first option controls how the takeoff is handled from planets; it either prompts you about the incomplete ships (default) or skips it and takes off straight away.

The second option handles how the fighters are deployed; they are either deployed with the rest of the fighters or not. This only affects the automatic deploy-everything kind of command, manual deployments are always allowed.
There is also a new message about ships that were not deployed due to player preferences.

All messages and texts are subject to change, and if somebody more talented comes along then I'd be happy to use whatever they suggest. If we do end up using these messages then the preferences menu will need to become wider, because the current option is so long.

Tests are coming as soon as Derpy or Zitchas approves the idea (or sooner if I get bored). I don't want to separate them into another PR because all of the actual logic is here and I don't want to deal with any merge conflicts if they can be avoided.

**Comment by Hurleveur** [comment]
There's never too many options.
That said, we could have both of those in one, named `Incomplete fighters behaviour`, the default option would be "ask", then "launch", and finally, "deploy".
If you want we can even add a final "deploy and launch" option, if you want to keep both separate, otherwise "deploy" could mean we "launch" as well. (though may as well add that fourth option).

However, currently I don't think this works, since nothing in Preferences has been modified.
In order to make that kind of 3-4 choice preferences I'm talking about, you can look at https://github.com/endless-sky/endless-sky/pull/6346 and inspire yourself from the changes made there in Preferences and PreferencesPanel. ~~don't look at the lambdas within lambdas I don't want to scare you~~

Also would be neat to see what the new preference panel looks like with the extra option, once you're done.

**Comment by tibetiroka** [comment]
Preferences stores true/false options in a map, so the class doesn't need to be changes; PreferencesPanel has the entry and it handles the possible options. You only need to edit Preferences if the new preference isn't a simple on-off option (like the one you proposed).

A view of the panel: 
image
(The text is currently a placeholder)

The new options are saved automatically in preferences.txt:
```
volume 0
(...)
"Damaged fighters retreat" 1
"Deploy incomplete fighters" 0
"Draw background haze" 1
(...)
"System map sends move orders" 1
"Take off with incomplete fighters" 1
"Turrets focus fire" 1
(...)
maximized 1
```

**Comment by Hurleveur** [comment]
Nice, you even happen to add it in the only category that still has space! Still I think one option for all of this behavior is best, as you can see we kinda lack space in that panel, and the less options there are the clearer it is. (one option with more toggle is thus best)

**Comment by tibetiroka** [comment]
Okay, so just to clarify: you want a 3-state toggle with these options:
- Ask before launch, don't deploy (`ask`)
- Don't ask before launch, don't deploy (`launch`)
- Don't ask before launch, deploy (`deploy`)

(Launch and deploy sound a bit similar, but that is the content reviewer's problem.)

**Comment by Hurleveur** [comment]
Yeah exactly though the names may change, I think `depart` would be clearer, instead of `launch`?
I'm not too good with names so that can change, but you got what I mean indeed.

**Comment by tibetiroka** [comment]
Once again, I'm asking for your review.

**Review by Hurleveur — CHANGES_REQUESTED** [review]

File: source/AI.cpp [hunk 12]
```diff
@@ -134,8 +134,19 @@ namespace {
...
+		auto isCandidate = [](const shared_ptr<Ship> &ship, bool manual) -> bool
 		{
+			// Check for incomplete ships based on player preferences
+			// Only applied to automatic deployments, allows the player to manually deploy
+			if(!manual && !ship->HasDeployOrder() && Preferences::IncompleteFighterUsage() != "deploy")
+			{
+				auto flightChecks = ship->FlightCheck();
+				if(!flightChecks.empty())
+					for(const auto &result : flightChecks)
+						if(result.back() == '!')
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/AI.cpp:146]: This is a copy-paste of already existing code:
https://github.com/endless-sky/endless-sky/blob/4f4eac959982bd45ebe82c63f008f8c07f4dd587/source/PlanetPanel.cpp#L247-L259

[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/AI.cpp:146]: Since we are using it at two places now, it might actually be better to have like a bool method that checks for this, so you are right about that

[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/AI.cpp:146]: A separate bool method would be an alternative indeed.

Still, I don't like the "!" method but I can live with it if it was previous code. The biggest problem I have with it is optimization, we are generating a vector with strings in it so we can iterate over it and ignore most of the strings anyway? May as well not add those strings in, in the first place, so we have less to add and check later.

[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/AI.cpp:146]: ~~I used to make that mistake as well; assuming the old code is empty of bugs, and has good syntax.~~

[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/AI.cpp:146]: I think the ! and ? are used to differentiate between errors and warnings in tooltip messages, so they would have to stay unless we want to rename everything and implement a whole new system. I guess I'll just add an extra option that only adds errors and not warnings.

[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/AI.cpp:146]: I had a look, and it seems like this is the only case where we only check for errors; everywhere else we use the warnings as well. Not sure that warrants a change.

[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/AI.cpp]: Yeah I'm not sure either tbh I was hoping we could fix it in both places using this but indeed it'd be out of the scope of this PR.
Still, I think a little extra bool and one less place using iffy != "!" code won't hurt.

**Review by Hurleveur — APPROVED** [review]
Seems good to me now, only missing the tests that were mentioned earlier.
I did not playtest it, that should probably be done as well.

(btw you don't have to request a review all the time, I see whenever you make a change, though I personally don't mind it, at least its clear you're done changing what I asked)

**Comment by tibetiroka** [comment]
Okay, then I'm moving onto tests and have roadrunner re-review the text for the option when I'm done, since that is their area.

**Review by petervdmeer — CHANGES_REQUESTED** [review]
Adding new preferences makes the feature very visible for new players, which I feel is also inappropriate. So that part is also not acceptable to me.

There are a lot of open PRs on major features, and I still feel that this is a very minor improvement for a very small subset of players (and it might not even be the correct improvement; I would first like to see how this PR compares to alternatives like just storing ship-hulls in cargo).

Discussions and code-reviews on PRs can take quite a bit of time, so I prefer that we spend the (usually limited) time we have for ES on more major functionality.

Just silently ignoring this PR is also not appropriate, and I don't want to discourage anyone from working on features for ES, but I also don't want to give false hope. So I added the label "unlikely" to reflect my feeling that we have "bigger fish to fry".

**Comment by Amazinite** [comment]
I'm in agreement with Peter on this one.
